standard/types.md

```diff
@@ -976,4 +1024,26 @@ Later in the method, the code checks that `s` is not a null reference. The null-
 >
 > In the previous example, the backing field for the `DisappearingProperty` is set to null when it is read. However, a compiler may assume that reading a property doesn’t change the null state of that expression. *end example*
 
+A compiler may use any expression that dereferences a variable, property, or event to set the null state to not null. If it were null, the dereference expression would have thrown a `NullReferenceException`:
```

Comment:
Do we want to add a note about extension methods, where calling an extension method *looks* like it dereferences a variable, but doesn't actually do so?